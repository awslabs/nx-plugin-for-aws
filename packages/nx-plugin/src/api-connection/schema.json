{
  "$schema": "https://json-schema.org/schema",
  "$id": "APIConnection",
  "title": "",
  "type": "object",
  "properties": {
    "sourceProject": {
      "type": "string",
      "description": "The source project which will call the API",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the source project which will call the API",
      "x-dropdown": "projects"
    },
    "targetProject": {
      "type": "string",
      "description": "The target project containing your API",
      "x-prompt": "Select the target project containing your API",
      "x-dropdown": "projects"
    }
  },
  "required": ["sourceProject", "targetProject"]
}
