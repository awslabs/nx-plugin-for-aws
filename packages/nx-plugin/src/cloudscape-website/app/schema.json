{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"],
  "examplesFile": "../../../docs/application-examples.md"
}
