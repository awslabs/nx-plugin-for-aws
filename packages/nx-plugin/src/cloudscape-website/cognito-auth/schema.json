{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    },
    {
      "command": "nx g app myapp --classComponent",
      "description": "Use class components instead of functional components"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "the root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "cognitoDomain": {
      "description": "The cognito domain prefix to use",
      "type": "string",
      "x-priority": "important",
      "x-prompt": "the cognito domain prefix to use i.e: https://{{cognitoDomain}}-123456789.auth.ap-southeast-2.amazoncognito.com"
    },
    "allowSignup": {
      "description": "Whether to allow self-signup",
      "type": "boolean",
      "default": false,
      "x-prompt": "Enable self signup?",
      "x-priority": "important"
    }
  },
  "required": ["project", "cognitoDomain"]
}
