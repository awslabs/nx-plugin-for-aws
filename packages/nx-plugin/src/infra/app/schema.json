{
  "$schema": "https://json-schema.org/schema",
  "$id": "Infra",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "ruleSet": {
      "type": "string",
      "enum": [
        "aws_prototyping",
        "cfn_nag",
        "hipaa",
        "nist_csf",
        "pci_dss_3_2_1",
        "well_architected_reliability",
        "well_architected_security"
      ],
      "x-priority": "important",
      "x-prompt": "What rule set do you want to validate your AWS resources with?",
      "description": "Rule set to validate your AWS resources with.",
      "default": "aws_prototyping"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"]
}
