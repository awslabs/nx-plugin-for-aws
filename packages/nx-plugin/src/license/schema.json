{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "License",
  "title": "Add LICENSE files, and configure source code license headers",
  "type": "object",
  "properties": {
    "license": {
      "type": "string",
      "description": "SPDX License Identifier",
      "enum": ["Apache-2.0", "MIT", "ASL"],
      "default": "Apache-2.0",
      "x-priority": "important"
    },
    "copyrightHolder": {
      "type": "string",
      "description": "Copyright holder name",
      "default": "Amazon.com, Inc. or its affiliates",
      "x-priority": "important"
    }
  },
  "required": []
}
