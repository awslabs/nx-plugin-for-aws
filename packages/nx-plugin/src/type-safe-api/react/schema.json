{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "type-safe-api-react",
  "title": "Integrate a Type Safe API with a React website",
  "type": "object",
  "properties": {
    "hooksLibraryProject": {
      "type": "string",
      "description": "The name of the Type Safe API generated react hooks library project",
      "x-prompt": "The react hooks library project for the Type Safe API to integrate",
      "x-dropdown": "projects"
    },
    "frontendProject": {
      "type": "string",
      "description": "The name of the React project to add the Type Safe API integration to",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "React application's project name.",
      "x-dropdown": "projects"
    },
    "auth": {
      "type": "string",
      "description": "",
      "x-prompt": "Select your API authentication strategy",
      "x-dropdown": ["IAM", "None"],
      "default": "IAM"
    }
  },
  "required": ["hooksLibraryProject", "frontendProject", "auth"]
}
