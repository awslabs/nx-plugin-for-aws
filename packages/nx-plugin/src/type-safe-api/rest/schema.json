{
  "$schema": "https://json-schema.org/schema",
  "$id": "TypeSafeApiModel",
  "title": "Create a Type Safe REST API",
  "description": "Create a Type Safe REST API",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the API.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What would you like your API to be called? e.g. MyApi",
      "x-priority": "important"
    },
    "modelLanguage": {
      "description": "The language you would like to model your api in.",
      "type": "string",
      "enum": ["openapi", "typespec"],
      "default": "typespec",
      "x-prompt": "What language would you like to use to model your API?",
      "x-priority": "important"
    },
    "infrastructureLanguage": {
      "description": "The language you would like to write your infrastructure in.",
      "type": "string",
      "enum": ["typescript"],
      "default": "typescript",
      "x-prompt": "What language are you using to write CDK infrastructure?",
      "x-priority": "important"
    },
    "handlerLanguages": {
      "description": "The languages you would like to write lambda handlers in.",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript"]
      },
      "default": ["typescript"],
      "x-prompt": "What languages (if any) would you like to write lambda handlers in?",
      "x-priority": "important"
    },
    "libraries": {
      "description": "The libraries to generate from your API",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript-react-hooks"]
      },
      "default": ["typescript-react-hooks"],
      "x-prompt": "What libraries would you like to generate?",
      "x-priority": "important"
    },
    "runtimeLanguages": {
      "description": "The languages you would like to write client or server code in, in addition to the infrastructure or handler languages.",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript"]
      },
      "default": [],
      "x-prompt": "What other languages (if any) would you like to write client or server code in?"
    },
    "directory": {
      "type": "string",
      "description": "Parent directory where the api model is placed.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the project in?"
    },
    "scope": {
      "type": "string",
      "description": "Scope for API packages, e.g. @my-company. If omitted this will be inferred",
      "pattern": "^@[a-z0-9-]+(?:/[a-z0-9-]+)?$",
      "x-prompt": "What namespace would you like your library to have? i.e: @myapi"
    },
    "subDirectory": {
      "type": "string",
      "description": "The sub directory the API is placed in. By default this is the API name.",
      "x-prompt": "Which sub directory do you want to create the API in? (By default this is the API name)"
    }
  },
  "required": ["name", "modelLanguage", "infrastructureLanguage"]
}