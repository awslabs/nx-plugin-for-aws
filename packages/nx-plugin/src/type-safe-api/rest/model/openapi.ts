/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { addProjectConfiguration, generateFiles, Tree } from "@nx/devkit";
import { InferredTypeSafeRestApiSchema } from "../schema";
import { join } from "path";

export const openApiRestModelGenerator = async (tree: Tree, options: InferredTypeSafeRestApiSchema) => {
  addProjectConfiguration(tree, options.model.fullyQualifiedName, {
    root: options.model.dir,
    name: options.model.fullyQualifiedName,
    metadata: {
      apiName: options.name,
    } as unknown,
    targets: {
      build: {
        executor: "nx:run-commands",
        cache: true,
        options: {
          commands: [
            `type-safe-api parse-openapi-spec --specPath ${join(options.model.dir, "src", "main.yaml")} --outputPath ${options.model.outputSpecPath}`,
          ],
          parallel: false,
        },
      },
    },
  });

  generateFiles(tree, join(__dirname, "..", "files", "model", "openapi"), join(options.model.dir), {
    ...options,
  });
};
