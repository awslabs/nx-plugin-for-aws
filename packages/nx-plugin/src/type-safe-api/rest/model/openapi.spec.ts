/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree } from '@nx/devkit';
import { openApiRestModelGenerator } from './openapi';
import { join } from 'path';
import { InferredTypeSafeRestApiSchema } from '../schema';

describe('type-safe-api rest openapi model generator', () => {
  let tree: Tree;
  let options: InferredTypeSafeRestApiSchema;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    options = {
      name: 'test-api',
      modelLanguage: 'openapi',
      infrastructureLanguage: 'typescript',
      handlerLanguages: ['typescript'],
      libraries: [],
      runtimeLanguages: ['typescript'],
      fullyQualifiedApiName: 'test-api',
      nameKebabCase: 'test-api',
      namePascalCase: 'TestApi',
      dir: 'apis/test-api',
      model: {
        dir: 'apis/test-api-model',
        fullyQualifiedName: 'test-api-model',
        outputSpecPath: 'apis/test-api-model/dist/spec.json'
      },
      runtime: {
        dir: 'apis/test-api-runtime',
        typescript: {
          dir: 'apis/test-api-runtime/typescript',
          fullyQualifiedName: 'test-api-runtime-typescript'
        }
      },
      infrastructure: {
        dir: 'apis/test-api-infrastructure',
        typescript: {
          dir: 'apis/test-api-infrastructure/typescript',
          fullyQualifiedName: 'test-api-infrastructure-typescript'
        }
      },
      handlers: {
        dir: 'apis/test-api-handlers',
        typescript: {
          dir: 'apis/test-api-handlers/typescript',
          fullyQualifiedName: 'test-api-handlers-typescript',
          assetPath: 'apis/test-api-handlers/typescript/dist'
        }
      },
      library: {
        dir: 'apis/test-api-library'
      }
    };
  });

  it('should create and configure project.json', async () => {
    await openApiRestModelGenerator(tree, options);

    // Check project.json was created
    expect(tree.exists('apis/test-api-model/project.json')).toBeTruthy();

    // Verify project.json contents
    const projectConfig = JSON.parse(
      tree.read('apis/test-api-model/project.json', 'utf-8')
    );

    // Check metadata
    expect(projectConfig.metadata.apiName).toBe('test-api');

    // Check build target configuration
    expect(projectConfig.targets.build).toBeDefined();
    expect(projectConfig.targets.build.executor).toBe('nx:run-commands');
    expect(projectConfig.targets.build.options.commands[0]).toContain('type-safe-api parse-openapi-spec');
    expect(projectConfig.targets.build.options.commands[0]).toContain(join('apis/test-api-model', 'src', 'main.yaml'));
    expect(projectConfig.targets.build.options.commands[0]).toContain(options.model.outputSpecPath);

    // Snapshot full project.json
    expect(projectConfig).toMatchSnapshot('project.json');
  });

  it('should generate OpenAPI model files', async () => {
    await openApiRestModelGenerator(tree, options);

    // Check template files were generated
    expect(tree.exists('apis/test-api-model/src/main.yaml')).toBeTruthy();
    expect(tree.read('apis/test-api-model/src/main.yaml', 'utf-8')).toMatchSnapshot(
      'main.yaml'
    );
  });
});
