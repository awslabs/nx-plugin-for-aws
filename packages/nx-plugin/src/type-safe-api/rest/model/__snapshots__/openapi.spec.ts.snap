// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type-safe-api rest openapi model generator > should create and configure project.json > project.json 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "metadata": {
    "apiName": "test-api",
  },
  "name": "test-api-model",
  "targets": {
    "build": {
      "cache": true,
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "type-safe-api parse-openapi-spec --specPath apis/test-api-model/src/main.yaml --outputPath apis/test-api-model/dist/spec.json",
        ],
        "parallel": false,
      },
    },
  },
}
`;

exports[`type-safe-api rest openapi model generator > should generate OpenAPI model files > main.yaml 1`] = `
"openapi: 3.0.0
info:
  title: "TestApi"
  version: "1.0"
  description: "TestApi"

paths:
  /hello:
    get:
      summary: Say hello
      operationId: SayHello
      x-handler:
        language: "typescript"
      parameters:
        - name: name
          in: query
          schema:
            type: string
          required: false
          description: Name to say hello to
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  message:
                    type: string
                required:
                  - message
        '400':
          description: Bad Request Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BadRequestErrorResponseContent'
        '403':
          description: Not Authorized Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/NotAuthorizedErrorResponseContent'
        '404':
          description: Not Found Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/NotFoundErrorResponseContent'
        '500':
          description: Internal Failure Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/InternalFailureErrorResponseContent'

components:
  schemas:
    BadRequestErrorResponseContent:
      type: object
      description: An error at the fault of the client sending invalid input
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    NotAuthorizedErrorResponseContent:
      type: object
      description: An error due to the client not being authorized to access the resource
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    NotFoundErrorResponseContent:
      type: object
      description: An error due to the client attempting to access a missing resource
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    InternalFailureErrorResponseContent:
      type: object
      description: An internal failure at the fault of the server
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message
"
`;
